File: samples/cext-trey-research-auth/src/services/AssignmentDbService.ts
import DbService from './DbService';
import { DbAssignment } from '../model/dbModel';
import { Assignment } from '../model/baseModel';
import { HttpError } from './Utilities';

const TABLE_NAME = "Assignment";

class AssignmentDbService {

    // NOTE: Assignments are READ-WRITE so disable local caching
    private dbService = new DbService<DbAssignment>(false);

    async getAssignments(): Promise<Assignment[]> {
        const assignments = await this.dbService.getEntities(TABLE_NAME) as DbAssignment[];
        const result = assignments.map((e) => this.convertDbAssignment(e));
        return result;
    }

    async chargeHoursToProject(projectId: string, consultantId: string, month: number, year: number, hours: number): Promise<number> {
        try {
            const dbAssignment = await this.dbService.getEntityByRowKey(TABLE_NAME, projectId + "," + consultantId) as DbAssignment;
            if (!dbAssignment) {
                throw new HttpError(404, "Assignment not found");
            }
            // Add the hours delivered
            if (!dbAssignment.delivered) {
                dbAssignment.delivered = [{ month: month, year: year, hours: hours }];
            } else {
                let a = dbAssignment.delivered.find(d => d.month === month && d.year === year);
                if (a) {
                    a.hours += hours;
                } else {
                    dbAssignment.delivered.push({ month, year, hours });
                }
            }
            dbAssignment.delivered.sort((a, b) => a.year - b.year || a.month - b.month);

            // Subtract the hours from the forecast
            let remainingForecast = -hours;
            if (!dbAssignment.forecast) {
                dbAssignment.forecast = [{ month: month, year: year, hours: -hours }];
            } else {
                let a = dbAssignment.forecast.find(d => d.month === month && d.year === year);
                if (a) {
                    a.hours -= hours;
                    remainingForecast = a.hours;
                } else {
                    dbAssignment.forecast.push({ month: month, year: year, hours: -hours });
                }
            }
            dbAssignment.forecast.sort((a, b) => a.year - b.year || a.month - b.month);

            await this.dbService.updateEntity(TABLE_NAME, dbAssignment)

            return remainingForecast;
        } catch (e) {
            throw new HttpError(404, "Assignment not found");
        }
    }

    async addConsultantToProject(projectId: string, consultantId: string, role: string, hours: number): Promise<number> {

        const month = new Date().getMonth() + 1;
        const year = new Date().getFullYear();

        let dbAssignment = null;
        try {
            dbAssignment = await this.dbService.getEntityByRowKey(TABLE_NAME, projectId + "," + consultantId) as DbAssignment;
        } catch { }

        if (dbAssignment) {
            throw new HttpError(403, "Assignment already exists");
        }

        try {
            const newAssignment: DbAssignment = {
                etag: "",
                partitionKey: TABLE_NAME,
                rowKey: projectId + "," + consultantId,
                timestamp: new Date(),
                id: projectId + "," + consultantId,
                projectId: projectId,
                consultantId: consultantId,
                role: role,
                billable: true,
                rate: 100,
                forecast: [{ month: month, year: year, hours: hours }],
                delivered: []
            };

            await this.dbService.createEntity(TABLE_NAME, newAssignment.id, newAssignment)

            return hours;
        } catch (e) {
            throw new HttpError(500, "Unable to add assignment");
        }
    }

    private convertDbAssignment(dbAssignment: DbAssignment): Assignment {
        const result: Assignment = {
            id: dbAssignment.id,
            projectId: dbAssignment.projectId,
            consultantId: dbAssignment.consultantId,
            role: dbAssignment.role,
            billable: dbAssignment.billable,
            rate: dbAssignment.rate,
            forecast: dbAssignment.forecast,
            delivered: dbAssignment.delivered
        };

        return result;
    }
}

export default new AssignmentDbService();

File: samples/cext-trey-research-python/services/assignment_db_service.py
from .db_service import DbService
from model.db_model import DbAssignment
from model.base_model import Assignment
from .utilities import HttpError
from datetime import datetime
import logging
TABLE_NAME = "Assignment"

class AssignmentDbService:

    # NOTE: Assignments are READ-WRITE so disable local caching
    def __init__(self):
        self.db_service = DbService[DbAssignment](ok_to_cache_locally=False)

    async def get_assignments(self) -> list[Assignment]:
        assignments = await self.db_service.get_entities(TABLE_NAME)
        result = [self.convert_db_assignment(e) for e in assignments]
        return result

    async def charge_hours_to_project(self, project_id: str, consultant_id: str, month: int, year: int, hours: int) -> int:
        try:
            logging.info(f"tablename : {TABLE_NAME},project id : {project_id}, consultant_id :{consultant_id} ")
            db_assignment = await self.db_service.get_entity_by_row_key(TABLE_NAME, f"{project_id},{consultant_id}")
            if not db_assignment:
                raise HttpError(404, "Assignment not found")

            # Add the hours delivered
            if not db_assignment.get("delivered"):
                db_assignment["delivered"] = [{"month": month, "year": year, "hours": hours}]
            else:
                assignment_record = next((d for d in db_assignment["delivered"] if d["month"] == month and d["year"] == year), None)
                if assignment_record:
                    assignment_record["hours"] += hours
                else:
                    db_assignment["delivered"].append({"month": month, "year": year, "hours": hours})

            db_assignment["delivered"].sort(key=lambda d: (d["year"], d["month"]))

            # Subtract the hours from the forecast
            remaining_forecast = -hours
            if not db_assignment.get("forecast"):
                db_assignment["forecast"] = [{"month": month, "year": year, "hours": -hours}]
            else:
                forecast_record = next((d for d in db_assignment["forecast"] if d["month"] == month and d["year"] == year), None)
                if forecast_record:
                    forecast_record["hours"] -= hours
                    remaining_forecast = forecast_record["hours"]
                else:
                    db_assignment["forecast"].append({"month": month, "year": year, "hours": -hours})

            db_assignment["forecast"].sort(key=lambda d: (d["year"], d["month"]))

            await self.db_service.update_entity(TABLE_NAME, db_assignment)
            return remaining_forecast

        except Exception as e:
            raise HttpError(404, f"Assignment not found : {e}")


    async def add_consultant_to_project(self, project_id: str, consultant_id: str, role: str, hours: int) -> int:
        month = datetime.utcnow().month
        year = datetime.utcnow().year

        db_assignment = None
        try:
            db_assignment = await self.db_service.get_entity_by_row_key(TABLE_NAME, f"{project_id},{consultant_id}")
        except Exception:
            pass

        if db_assignment:
            raise HttpError(403, "Assignment already exists")

        try:
            new_assignment = DbAssignment(
                partition_key=TABLE_NAME,
                row_key=f"{project_id},{consultant_id}",
                timestamp=datetime.utcnow(),
                id=f"{project_id},{consultant_id}",
                projectId=project_id,
                consultantId=consultant_id,
                role=role,
                billable=True,
                rate=100,
                forecast=[{"month": month, "year": year, "hours": hours}],
                delivered=[],
                etag=""
            )

            await self.db_service.create_entity(TABLE_NAME, new_assignment.row_key, new_assignment)
            return hours

        except Exception as e:
            raise HttpError(500, f"Unable to add assignment :{e}")


    def convert_db_assignment(self, db_assignment: DbAssignment) -> Assignment:
        result = Assignment(
            id=db_assignment.get('id'),
            projectId=db_assignment.get('projectId'),
            consultantId=db_assignment.get('consultantId'),
            role=db_assignment.get('role'),
            billable=db_assignment.get('billable'),
            rate=db_assignment.get('rate'),
            forecast=db_assignment.get('forecast'),
            delivered=db_assignment.get('delivered')
        )
        return result

# Instantiate and use the AssignmentDbService
assignment_db_service = AssignmentDbService()